LogMaker/FunctionTracer/call_cust.cpp

```diff
@@ -1,175 +1,133 @@
-{
-    if( !taken ) return;
-    
-    const string *s = Target2String(target);
-    do_call_args(s, arg0);
-
-    if (s != &invalid)
-        delete s;
-}
-
-/* ===================================================================== */
-
-VOID  do_call(const string *s)
-{
+VOID do_call(const string *s,ADDRINT target)
+{   
     // Here we check for _Exit because that's the last call happening before the binary teminates.
-    if(*s != "_Exit")
-        calls_list.push_back(*s);
-    else if(*s == "_Exit"){ 
-        for(unsigned int i = 0; i < calls_list.size();i++)
-        {
-            string ser = calls_list[i];
-            //Switch cases for parsing the list we have 
-            switch (ser[0])
-            {
-            case '.':
-                break;
-            case '_':
-                break;
-            case '$':
-                break;
-            case '(':
-                break;
-            case '#':
-                break;
-            case '@':
-                break;
-            default:
-                //cout << ser[0] << endl;
-                //TraceFile << ser << endl;
-                auto result = countMap.insert(std::pair<std::string, int>(ser, 1));
-	            if (result.second == false)
-		            result.first->second++;
-                break;
+    if (s->rfind("__libc_start_main") == 0){
+        isMain = 1;
+    }
+    if ((s->rfind("exit") == 0) || s->rfind("_exit") == 0){
+        isExit = 0;
+    }
+    if(isMain && isExit){
+        if(target > LIBC_BASE){
+            isLib = 1;
+        }
+        else {
+            isLib = 0;
+        }
+        for (auto &itr : Func_list){
+            if (itr.name == s->c_str()){
+                itr.count++;
+                return ;
             }
         }
-        if ( std::find(calls_list.begin(), calls_list.end(), "ptrace@plt") != calls_list.end() ) 
+        switch (s->c_str()[0])
         {
-            cout << "================Analysis Complete================"<<endl;
-            cout << "PTRACE detected" <<endl;
-            cout << "======================Stats ====================="<<endl;
-        }
-        for (auto & elem : countMap){
-	    // If frequency count is greater than 1 then its a duplicate element
-	        if (elem.second >= 1){
-		        std::cout << elem.first << " was called " << elem.second << " times."<< std::endl;
-	        }
+        case '.':
+            break;
+        case '_':
+            break;
+        case '$':
+            break;
+        case '(':
+            break;
+        case '#':
+            break;
+        case '@':
+            break;
+        default:
+            Func_list.push_back({s->c_str(),(target - image_base),1,isLib});
         }
     }
 }
 
-
-/* ===================================================================== */
-
 VOID  do_call_indirect(ADDRINT target, BOOL taken)
 {
     if( !taken ) return;
 
     const string *s = Target2String(target);
```

Comment:
> Currently, the issue is that the offsets of those library function are negative as the address is before the image_base

how are the correct addresses showing up in the function names? can't we use the same address, or is offset something else? @pranjalsingh008

Reply:
As mentioned in the above comment we can store all the offsets of the images and then calculate this offset. But then do we really need to provide statistics of these functions ?